File: 4_Behavioral_Design_Patterns/4.01_Chain_Of_Responsibility/Python_Snippet/COR.py
class Expense:
    def __init__(self, amount):
        self.amount = amount

class Approver:
    def __init__(self, name, approval_limit, next_approver=None):
        self.name = name
        self.approval_limit = approval_limit
        self.next_approver = next_approver

    def approve_expense(self, expense):
        if expense.amount <= self.approval_limit:
            print(f"{self.name} approved the expense of ${expense.amount}")
        elif self.next_approver:
            print(f"{self.name} cannot approve. Passing to {self.next_approver.name}.")
            self.next_approver.approve_expense(expense)
        else:
            print(f"None of the approvers can handle this expense of ${expense.amount}")

# Example usage
manager3 = Approver("Manager 3", 300)
manager2 = Approver("Manager 2", 200, manager3)
manager1 = Approver("Manager 1", 100, manager2)

expense1 = Expense(150)
expense2 = Expense(250)
expense3 = Expense(350)

manager1.approve_expense(expense1)
print("-------------------")
manager1.approve_expense(expense2)
print("-------------------")
manager1.approve_expense(expense3)

File: 4_Behavioral_Design_Patterns/4.01_Chain_Of_Responsibility/Java_Snippet/COR.java
class Expense {
    double amount;

    public Expense(double amount) {
        this.amount = amount;
    }
}

class Approver {
    String name;
    double approvalLimit;
    Approver nextApprover;

    public Approver(String name, double approvalLimit, Approver nextApprover) {
        this.name = name;
        this.approvalLimit = approvalLimit;
        this.nextApprover = nextApprover;
    }

    public void approveExpense(Expense expense) {
        if (expense.amount <= approvalLimit) {
            System.out.println(this.name + " approved the expense of " + expense.amount);
        } else if (nextApprover != null) {
            System.out.println(this.name + " cant approve. Passing to " + nextApprover.name);
            nextApprover.approveExpense(expense);
        } else {
            System.out.println("None of the approvers can approve the expense of " + expense.amount);
        }
    }
}

public class COR {
    public static void main(String[] args) {
        Approver manager3 = new Approver("Manager 3", 300, null);
        Approver manager2 = new Approver("Manager 2", 200, manager3);
        Approver manager1 = new Approver("Manager 1", 100, manager2);

        Expense expense1 = new Expense(150);
        Expense expense2 = new Expense(250);
        Expense expense3 = new Expense(350);

        manager1.approveExpense(expense1);
        System.out.println("-------------------");
        manager1.approveExpense(expense2);
        System.out.println("-------------------");
        manager1.approveExpense(expense3);
    }
}